test/test_spectral_ops.py

```diff
@@ -43,57 +47,229 @@ def method_fn(t):
+            torch.complex128: torch.complex128,
+        }
+        T = torch.fft.fft(t)
+        self.assertEqual(T.dtype, PROMOTION_MAP[dtype])
+
+        PROMOTION_MAP_C2R = {
+            torch.int8: torch.float,
+            torch.float: torch.float,
+            torch.double: torch.double,
+            torch.complex64: torch.float,
+            torch.complex128: torch.double,
+        }
+        R = torch.fft.hfft(t)
+        self.assertEqual(R.dtype, PROMOTION_MAP_C2R[dtype])
+
+        if not dtype.is_complex:
+            PROMOTION_MAP_R2C = {
+                torch.int8: torch.complex64,
+                torch.float: torch.complex64,
+                torch.double: torch.complex128,
+            }
+            C = torch.fft.rfft(t)
+            self.assertEqual(C.dtype, PROMOTION_MAP_R2C[dtype])
+
+    @skipCUDAIfRocm
+    @skipCPUIfNoMkl
+    @onlyOnCPUAndCUDA
+    @dtypes(torch.half, torch.bfloat16)
+    def test_fft_half_errors(self, device, dtype):
+        # TODO: Remove torch.half error when complex32 is fully implemented
+        x = torch.randn(64, device=device).to(dtype)
+        fft_functions = (torch.fft.fft, torch.fft.ifft,
+                         torch.fft.rfft, torch.fft.irfft,
+                         torch.fft.hfft, torch.fft.ihfft)
+        for fn in fft_functions:
+            with self.assertRaisesRegex(RuntimeError, "Unsupported dtype "):
+                fn(x)
+
+    @skipCPUIfNoMkl
+    @skipCUDAIfRocm
+    @dtypes(torch.double, torch.complex128)  # gradcheck requires double
+    def test_fft_backward(self, device, dtype):
+        test_args = list(product(
+            # input
+            (torch.randn(67, device=device, dtype=dtype),
+             torch.randn(9, 6, 3, device=device, dtype=dtype)),
+            # n
+            (None, 6),
+            # dim
+            (-1, 0),
+            # norm
+            (None, "forward", "backward", "ortho")
+        ))
+
+        fft_functions = ['fft', 'ifft', 'hfft', 'irfft']
+        # Real-only functions
+        if not dtype.is_complex:
+            fft_functions += ['rfft', 'ihfft']
+
+        for fname in fft_functions:
+            torch_fn = getattr(torch.fft, fname)
+
+            for iargs in test_args:
+                args = list(iargs)
+                input = args[0]
+                args = args[1:]
+
+                # Workaround for gradcheck's poor support for complex input
+                # Use real input instead and put view_as_complex into the graph
+                if dtype.is_complex:
+                    def test_fn(x):
+                        return torch_fn(torch.view_as_complex(x), *args)
+                    input = torch.view_as_real(input).detach().requires_grad_()
+                else:
+                    def test_fn(x):
+                        return torch_fn(x, *args)
+                    input = input.detach().requires_grad_()
+
+                self.assertTrue(torch.autograd.gradcheck(test_fn, (input,)))
```

Comment:
Just to draw attention to this. I was getting `gradcheck` failures for `irfft` but when I looked into it, the numerical jacobian calculated inside of `gradcheck` was actually wrong and tweaking the `eps` didn't help. `torch.irfft` however worked fine, and I found that making the leaf node purely real fixed the numerical jacobian issue.

Reply:
Interesting. cc @albanD and @anjali411, who are working on complex autograd.

Reply:
Yes there is not support for complex in gradcheck yet. So only running it in the real case is enough right now.